rust/src/graph/higher_order_queries.rs

```diff
@@ -208,3 +208,80 @@ impl Graph {
         )
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::graph::Graph;
+    use rustc_hash::FxHashSet;
+
+    #[test]
+    fn test_generate_module_permutations_simple_layers() {
+        let mut graph = Graph::default();
+
+        let top_module = graph.get_or_add_module("app.top").token;
+        let middle_module = graph.get_or_add_module("app.middle").token;
+        let bottom_module = graph.get_or_add_module("app.bottom").token;
+
+        let mut top_layer = FxHashSet::default();
+        top_layer.insert(top_module);
+
+        let mut middle_layer = FxHashSet::default();
+        middle_layer.insert(middle_module);
+
+        let mut bottom_layer = FxHashSet::default();
+        bottom_layer.insert(bottom_module);
+
+        let top_level = Level::new(top_layer, false);
+        let middle_level = Level::new(middle_layer, false);
+        let bottom_level = Level::new(bottom_layer, false);
+
+        let levels = vec![top_level, middle_level, bottom_level];
+
+        let permutations = graph.generate_module_permutations(&levels);
```

Comment:
Coming to this function afresh, makes me think it's not very well named (or at least could do with more documentation) - do you agree?